### PR TITLE
Mix Dependency Improvements

### PR DESCRIPTION
**Pull request by scrogson**
* Add mix.lock to git
* Set deps environment groups inline
* Make `ibrowse` optional
* Use an empty list as default option for `with` macro

[Review on Reviewable](https://reviewable.io/reviews/monterail/tesla/3)

### CODE REVIEW

File: lib/tesla/adapter/ibrowse.ex [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+          gather_response(env, target, {:ok, _status, _headers, body})
+
+        {:ibrowse_async_response_end, _} ->
+          response = format_response(env, res)
+          send target, {:tesla_response, response}
+      end
+    end
+  end
+end
```
Comment (teamon) [inline lib/tesla/adapter/ibrowse.ex]: Does `Code.ensure_loaded?(:ibrowse)` work in compile or run-time?

File: mix.exs [hunk 2]
```diff
@@ -30,18 +30,10 @@ defmodule Tesla.Mixfile do
               "Documentation" => "http://hexdocs.pm/tesla/0.1.0/"}]
   end
 
-  defp deps(:dev) do
-    [{:ibrowse, github: "cmullaparthi/ibrowse", tag: "v4.1.1"},
+  defp deps do
+    [{:ibrowse, github: "cmullaparthi/ibrowse", tag: "v4.1.1", optional: true},
```
Comment (teamon) [inline mix.exs]: How does `optional: true` work?

Comment (teamon) [inline mix.exs]: Also I think `exjsx` should not be a hard dependency.

**Comment by teamon** [comment]
Wow, thanks a lot!

I've already merged #2 before seeing this one so I guess a rebase will be needed.

**Comment by scrogson** [comment]
@teamon I've rebased this one on master.

File: mix.exs (hunk 2 above)
Comment (scrogson) [inline mix.exs]: `optional: true` sets the dependency is optional and used only to specify requirements. It is left to the user to include the dependency only if they wish to use it.

This is the same way dependencies are setup in `Plug` and `Ecto`. `Ecto` has an optional dependency on `postgrex` and `mariaex` and it is required for the user to include the dependency they wish to use.

I think `exjsx` should eventually not be a dependency at all. A JSON encoder/decoder should be left to the user to include and specify when adding the middleware (Similar to how `Plug` allows the JSON Parser to be configured).

**Pull request merged**